apps/prairielearn/assets/scripts/lib/questionsTable.js

```diff
@@ -0,0 +1,194 @@
+/* eslint-env browser, jquery */
+
+import _ from 'lodash';
+import { onDocumentReady, decodeData } from '@prairielearn/browser-utils';
+import { html } from '@prairielearn/html';
+
+onDocumentReady(() => {
+  const { course_instance_ids, showAddQuestionButton, urlPrefix, plainUrlPrefix } =
+    decodeData('questions-table-data');
+  window.topicList = function () {
+    var data = $('#questionsTable').bootstrapTable('getData');
+    return _.keyBy(_.map(data, (row) => row.topic.name));
+  };
+
+  window.tagsList = function () {
+    var data = $('#questionsTable').bootstrapTable('getData');
+    return _.keyBy(_.map(_.flatten(_.filter(_.map(data, (row) => row.tags))), (row) => row.name));
+  };
+
+  window.versionList = function () {
+    var data = $('#questionsTable').bootstrapTable('getData');
+    return _.keyBy(_.map(data, (row) => row.display_type));
+  };
```

Comment:
I'm wondering if it may be better to have the bootstrap-table data be set in JS instead of via data attributes, particularly the ones referring to functions. An example: https://github.com/PrairieLearn/PrairieLearn/blob/cfbab7a69b683896deba405e3a581e08a3813908/apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.ejs#L18-L184.

Reply:
To clarify: I think the table data itself (`data-data`) could still remain in the HTML, I'm referring to the other settings.

Reply:
If it's possible to pass the data itself via JS then I'd also prefer that. That is, we send the data from server->client using `encodedData()` and then the page JS gives it to the table.

Reply:
There is another option to pass the data to the table, which is to use a separate URL that includes the data in JSON format. This has been used in other pages, like [here](https://github.com/PrairieLearn/PrairieLearn/blob/b1bebf266885b1edd39cf64bfea53a30a5a4c708/apps/prairielearn/src/pages/instructorGradebook/instructorGradebook.ejs#L14), [here](https://github.com/PrairieLearn/PrairieLearn/blob/b1bebf266885b1edd39cf64bfea53a30a5a4c708/apps/prairielearn/src/pages/instructorAssessmentInstances/instructorAssessmentInstances.ejs#L198) and [here](https://github.com/PrairieLearn/PrairieLearn/blob/b1bebf266885b1edd39cf64bfea53a30a5a4c708/apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.ejs#L20). This of course requires a separate end point for the data, but it has a few advantages, like loading the page faster while waiting for a large amount of data, and the ability to reload the table data without refreshing the page (which also means filters, etc. are not lost on reload).